README.md
# Neural Coding Lab Website

forked from janczizikow/sleek

## License

The theme is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


news.md
---
layout: news
title: News
permalink: /news/
---



makeImages.m


function makeImages(sourceImg)

% load image data
%%
img = imread(sourceImg); 

% is it too wide or too tall? 
desRatio = 1920/1080; 
thisRatio = size(img,2)/size(img,1); 

% will fill with background color determined as the mode of all edge pixels
bckgcolor = mode([squeeze(img(1,:,:)); ...
    squeeze(img(end,:,:));...
    squeeze(permute(img(:,1,:),[2 1 3])); ...
    squeeze(permute(img(:,end,:),[2 1 3]))],1);
bckgcolor = permute(bckgcolor, [1 3 2]);
if thisRatio>desRatio % too wide, need to add rows
    desRows = ceil(size(img,2)/desRatio); 
    tempimg = repmat(bckgcolor, desRows, size(img,2), 1);
    idx = floor((desRows-size(img,1))/2);
    tempimg(idx:idx+size(img,1)-1,:,:) = img; 
    img = tempimg;
elseif thisRatio<desRatio % too tall, need to add columns
    desCols = ceil(size(img,1)*desRatio);
    tempimg = repmat(bckgcolor, size(img,1), desCols, 1);
    idx = floor((desCols-size(img,2))/2);
    tempimg(:,idx:idx+size(img,2)-1,:) = img; 
    img = tempimg;
end

%%
[d,imName] = fileparts(sourceImg);
[~,dest] = fileparts(d);

outputDir = fullfile('assets', 'img', dest); 

targets = struct(); 
targets(end).width = 230; targets(end).suffix = 'placehold';
targets(end+1).width = 535; targets(end).suffix = 'thumb';
targets(end+1).width = 535*2; targets(end).suffix = 'thumb@2x';
targets(end+1).width = 575; targets(end).suffix = 'xs';
targets(end+1).width = 767; targets(end).suffix = 'sm';
targets(end+1).width = 991; targets(end).suffix = 'md';
targets(end+1).width = 1999; targets(end).suffix = 'lg';
targets(end+1).width = 1920; targets(end).suffix = '';

qual = 70; 
%%

origW = size(img,2); 
for ii = 1:numel(targets)
    if ~isempty(targets(ii).suffix)
        fn = [imName '_' targets(ii).suffix '.jpg']; 
    else 
        fn = [imName '.jpg']; 
    end
    imgR = imresize(img, targets(ii).width/origW); 
    imwrite(imgR, fullfile(outputDir,fn), 'JPEG', 'Quality', qual);
end

repAuthors.m

%%

repRoot = '/Users/nick/github/Steinmetzlab.github.io/';
q = loadjson(fullfile(repRoot, '_data/oldPubs.json'));

%%

for n = 1:numel(q.papers)
    
    s = q.papers{n}.authors; 
    ss = strsplit(strrep(strrep(s, '<b>', ''), '</b>', ''), ', ');
    
    noInits = cellfun(@(x)strsplit(x,' '), ss, 'uni', false);
    for x = 1:numel(noInits)
        if strcmp(noInits{x}{1}, 'et')
            noInits{x} = 'et al.'; 
        elseif numel(noInits{x})>1 && contains(noInits{x}{2}, '*')
            noInits{x} = [noInits{x}{1} '*'];
        else
            noInits{x} = noInits{x}{1};
        end
    end
%     noInits = cellfun(@(x)x{1}, noInits, 'uni', false);
    
    q.papers{n}.authors = ss; 
    q.papers{n}.authors_short = noInits;
    
    % also fix tags??
    
    q.papers{n}.tags = cellfun(@(x)x{1}, q.papers{n}.tags, 'uni', false);
    
end

%%

j = savejson('', q);

j = strrep(j, '\/', '/'); % wtf

fid = fopen(fullfile(repRoot, '_data/pubs.json'), 'w');
fwrite(fid, j, 'char');

gulpfile.js
'use strict';

// Gulp and node
const gulp = require('gulp');
const cp = require('child_process');
const notify = require('gulp-notify');
const size = require('gulp-size');

// Basic workflow plugins
const browserSync = require('browser-sync');
const browserify = require('browserify');
const source = require('vinyl-source-stream');
const buffer = require('vinyl-buffer');
const clean = require('gulp-clean');
const sass = require('gulp-sass');
const jekyll = process.platform === 'win32' ? 'jekyll.bat' : 'jekyll';
const messages = {
    jekyllBuild: '<span style="color: grey">Running:</span> $ jekyll build'
};

// Performance workflow plugins
const htmlmin = require('gulp-htmlmin');
const prefix = require('gulp-autoprefixer');
const sourcemaps = require('gulp-sourcemaps');
const uglify = require('gulp-uglify');
const critical = require('critical');
const sw = require('sw-precache');

// Image Generation
const responsive = require('gulp-responsive');
const $ = require('gulp-load-plugins')();
const rename = require('gulp-rename');
const imagemin = require('gulp-imagemin');

const src = {
  css: '_sass/jekyll-sleek.scss',
  js: '_js/scripts.js',
}
const dist = {
  css: '_site/assets/css',
  js: '_site/assets/js',
}

function handleErrors() {
  var args = Array.prototype.slice.call(arguments);
  notify.onError({
    title: 'Compile Error',
    message: '<%= error.message %>'
  }).apply(this, args);
  this.emit('end'); // Keep gulp from hanging on this task
}

// Build the Jekyll Site
gulp.task('jekyll-build', function (done) {
    browserSync.notify(messages.jekyllBuild);
    return cp.spawn( jekyll , ['build'], {stdio: 'inherit'})
        .on('close', done);
});

gulp.task('deploy', ['jekyll-build'], function () {
    return gulp.src('./_site/**/*')
        .pipe(deploy());
});

// Rebuild Jekyll & do page reload
gulp.task('rebuild', ['jekyll-build'], function (done) {
    browserSync.reload();
    done();
});

// Serve after jekyll-build
gulp.task('browser-sync', ['sass', 'js', 'sw', 'jekyll-build'], function() {
    browserSync({
        server: {
            baseDir: '_site'
        }
    });
});

// SASS
gulp.task('sass', function() {
  return gulp.src(src.css)
    .pipe(sourcemaps.init())
    .pipe(sass({
      outputStyle: 'compressed',
      includePaths: ['scss'],
      onError: browserSync.notify
    }).on('error', sass.logError))
    .pipe(sourcemaps.write({includeContent: false}))
    .pipe(sourcemaps.init({loadMaps: true}))
    .pipe(prefix())
    .pipe(sourcemaps.write('./'))
    .pipe(rename({ basename: 'main' }))
    .pipe(gulp.dest(dist.css))
    .pipe(browserSync.reload({ stream: true }))
    .pipe(gulp.dest('assets/css'));
});

//  JS
gulp.task('js', function() {
  return browserify(src.js, {debug: true, extensions: ['es6']})
    .transform('babelify', {presets: ['es2015']})
    .bundle()
    .on('error', handleErrors)
    .pipe(source('bundle.js'))
    .pipe(buffer())
    .pipe(sourcemaps.init({loadMaps: true}))
    .pipe(uglify())
    .pipe(sourcemaps.write('./maps'))
    .pipe(size())
    .pipe(gulp.dest(dist.js))
    .pipe(browserSync.reload({stream: true}))
    .pipe(gulp.dest('assets/js'))
});

gulp.task('critical', function (cb) {
  critical.generate({
    base: '_site/',
    src: 'index.html',
    css: ['assets/css/main.css'],
    dimensions: [{
      width: 320,
      height: 480
    },{
      width: 768,
      height: 1024
    },{
      width: 1280,
      height: 960
    }],
    dest: '../_includes/critical.css',
    minify: true,
    extract: false,
    ignore: ['@font-face']
  });
});

gulp.task('watch', function() {
  gulp.watch('_sass/**/*.scss', ['sass']);
  gulp.watch(['*.html', '_layouts/*.html', '_includes/*.html', '_posts/*.md',  'pages_/*.md', '_include/*html'], ['rebuild']);
  gulp.watch(src.js, ['js']);
});

gulp.task('default', ['browser-sync', 'watch']);

// Minify HTML
gulp.task('html', function() {
    gulp.src('./_site/index.html')
      .pipe(htmlmin({ collapseWhitespace: true }))
      .pipe(gulp.dest('./_site'))
    gulp.src('./_site/*/*html')
      .pipe(htmlmin({ collapseWhitespace: true }))
      .pipe(gulp.dest('./_site/./'))
});

gulp.task('sw', function() {
  const rootDir ='./';
  const distDir = './_site';

  sw.write(`${rootDir}/sw.js`, {
    staticFileGlobs: [distDir + '/**/*.{js,html,css,png,jpg,svg}'],
    stripPrefix: distDir
  });
});

// Images
gulp.task('img', function() {
  return gulp.src('_img/posts/*.{png,jpg}')
    .pipe($.responsive({
      // For all the images in the folder
      '*': [{
        width: 230,
        rename: {suffix: '_placehold'},
      }, {
        // thubmnail
        width: 535,
        rename: { suffix: '_thumb' },
      }, {
        // thumbnail @2x
        width: 535 * 2,
        rename: { suffix: '_thumb@2x' },
      }, {
        width: 575,
        rename: { suffix: '_xs'}
      }, {
        width: 767,
        rename: {suffix: '_sm'}
      }, {
        width: 991,
        rename: { suffix: '_md' }
      }, {
        width: 1999,
        rename: { suffix: '_lg' }
      }, {
        // max-width hero
        width: 1920,
      }],
    }, {
      quality: 70,
      progressive: true,
      withMetadata: false,
      errorOnEnlargement: false,
    }))
    .pipe(imagemin())
    .pipe(gulp.dest('assets/img/posts/'));
});

gulp.task('img_people', function() {
  return gulp.src('_img/people/*.{png,jpg}')
    .pipe($.responsive({
      // For all the images in the folder
      '*': [{
        width: 230,
        rename: {suffix: '_placehold'},
      }, {
        // thubmnail
        width: 535,
        rename: { suffix: '_thumb' },
      }, {
        // thumbnail @2x
        width: 535 * 2,
        rename: { suffix: '_thumb@2x' },
      }, {
        width: 575,
        rename: { suffix: '_xs'}
      }, {
        width: 767,
        rename: {suffix: '_sm'}
      }, {
        width: 991,
        rename: { suffix: '_md' }
      }, {
        width: 1999,
        rename: { suffix: '_lg' }
      }, {
        // max-width hero
        width: 1920,
      }],
    }, {
      quality: 70,
      progressive: true,
      withMetadata: false,
      errorOnEnlargement: false,
    }))
    .pipe(imagemin())
    .pipe(gulp.dest('assets/img/people/'));
});

gulp.task('img_research', function() {
  return gulp.src('_img/research/*.{png,jpg}')
    .pipe($.responsive({
      // For all the images in the folder
      '*': [{
        width: 230,
        rename: {suffix: '_placehold'},
      }, {
        // thubmnail
        width: 535,
        rename: { suffix: '_thumb' },
      }, {
        // thumbnail @2x
        width: 535 * 2,
        rename: { suffix: '_thumb@2x' },
      }, {
        width: 575,
        rename: { suffix: '_xs'}
      }, {
        width: 767,
        rename: {suffix: '_sm'}
      }, {
        width: 991,
        rename: { suffix: '_md' }
      }, {
        width: 1999,
        rename: { suffix: '_lg' }
      }, {
        // max-width hero
        width: 1920,
      }],
    }, {
      quality: 70,
      progressive: true,
      withMetadata: false,
      errorOnEnlargement: false,
    }))
    .pipe(imagemin())
    .pipe(gulp.dest('assets/img/research/'));
});


gulp.task('clean', function () {
    return gulp.src('_site', {read: false})
      .pipe(clean());
});

gulp.task('serve', function() {
  browserSync({
    server: {
      baseDir: '_site'
    }
  });
});

gulp.task('build', ['sass', 'js', 'jekyll-build', 'img', 'sw']);
